File: 2023-Python/day-02/code.R
input <- readLines("2023-Python/day-02/input.txt") 

str_extract <- function(x, pattern, default = "0") {
  regmatches(x, regexec(pattern, x, perl = TRUE)) |> 
    lapply(\(x) if (length(x) == 0L) default else x) |> 
    unlist()
}

# -- Parsing -------------------------------------------------------------------
input_parsed <- input |> 
  lapply(function(line) {
    parts <- strsplit(line, ": ")[[1]]
    sets  <- strsplit(parts[2], "; ")[[1]]
    data.frame(
      game_id = as.integer(str_extract(parts[1], "(?<=Game )\\d+")),
      n_red   = as.integer(str_extract(sets, "\\d+(?= red)")), 
      n_green = as.integer(str_extract(sets, "\\d+(?= green)")),
      n_blue  = as.integer(str_extract(sets, "\\d+(?= blue)"))
    )
  }) |> 
  do.call(args = _, rbind)

max_cubes <- input_parsed |> 
  aggregate(by = list(input_parsed$game_id), max) 

# -- Part 1 --------------------------------------------------------------------
with(max_cubes, sum(unique(game_id[n_red < 12 & n_green < 13 & n_blue < 14])))

# -- Part 2 --------------------------------------------------------------------
with(max_cubes, sum(n_red * n_green * n_blue))

File: 2023-Python/day-02/code.py
import re

with open("2023-Python/day-02/input.txt") as input_file:
    input = input_file.readlines()

# -- Parsing -----------------------------------------------------------------------
def parse_line(line):
    line = line.split(": ")
    return {
        "ID": int(re.findall(r"\d+", line[0])[0]),
        "Sets": [parse_set(set) for set in line[1].split("; ")]
    }

def parse_set(set):
    return {
        re.findall(r"(red|green|blue)", turn)[0]: int(re.findall(r"\d+", turn)[0])
        for turn in set.split(", ")
    }
    
input_parsed = [parse_line(line) for line in input]

# -- Part 1 -----------------------------------------------------------------------
def is_possible(sets, reds, greens, blues):
    for set in sets:
        if set.get("red"  , 0) > reds  : return False
        if set.get("green", 0) > greens: return False
        if set.get("blue" , 0) > blues : return False
    return True

result1 = sum([
    game["ID"] 
    for game in input_parsed 
    if is_possible(game["Sets"], reds=12, greens=13, blues=14)
])

# -- Part 2 -----------------------------------------------------------------------
def min_set_power(sets):
    min_set = {"red": 0, "green": 0, "blue": 0}
    for set in sets:
        for colour, n in set.items():
            min_set.update({colour: max(min_set[colour], n)})
    return min_set["red"] * min_set["green"] * min_set["blue"]

result2 = sum([min_set_power(game["Sets"]) for game in input_parsed])
